File: 1-Beginner/sorting_algorithm_visualizer/mergesort.py
#merge sort algorithm for-loop ~kinda~
import time

# main algorithm
def mergesort(data,sleeper,drawdata):
    merge_sort_algorithm(data,0,len(data)-1,drawdata,sleeper)


def merge_sort_algorithm(data,left,right,drawdata,sleeper):
    if left < right:
        mid = (left+right)//2
        merge_sort_algorithm(data,left,mid,drawdata,sleeper)
        merge_sort_algorithm(data,mid+1,right,drawdata,sleeper)
        merge(data,left,mid,right,drawdata,sleeper)
        

def merge(data,left,mid,right,drawdata,sleeper):
    drawdata(data, colorarray(len(data),left,mid,right))
    time.sleep(sleeper)
    leftpart=data[left:mid+1]
    rightpart = data[mid+1:right+1]

    leftindex = rightindex = 0

    for dataindex  in range(left,right+1):
        if len(leftpart) > leftindex and len(rightpart) > rightindex:
            if leftpart[leftindex] >  rightpart[rightindex]:
                data[dataindex] = rightpart[rightindex]
                rightindex += 1
            else:
                data[dataindex] = leftpart[leftindex]
                leftindex += 1
        elif len(leftpart) > leftindex:
            data[dataindex] = leftpart[leftindex]
            leftindex += 1
        else:
            data[dataindex] = rightpart[rightindex]
            rightindex += 1

    drawdata(data,["pink" if x >= left and x <= right else "blue" for x in range(len(data))])
    time.sleep(sleeper)


def colorarray(n,left,mid,right):
    colorarr=[]
    for i in range(n):
        if i >= left and i <= right:
            if i>=left and i <= mid:
                colorarr.append("yellow")
            else:
                colorarr.append("green")
        else:
            colorarr.append("blue")
    return colorarr

File: 4-Visualization/sorting_algorithm_visualizer/C/mergesort.c
#include <stdio.h>

void merge(int arr[],int left,int mid,int right)
{
  
  
  int leftpart[(mid+1)-left];
  int rightpart[(right+1)-(mid+1)];
  
  // initializing left  array
  
  for (int i = left;i < mid + 1;i++)
  {
    leftpart[i] = arr[i];
  }

  // initializing right array
  for (int i = mid+1;i < right + 1;i++)
  {
    rightpart[i] = arr[i];
  }

  //size of leftpart
  int s_left_part = sizeof(leftpart)/sizeof(leftpart[0]); 
  //size of rightpart
  int s_right_part = sizeof(rightpart)/sizeof(rightpart[0]); 

  int leftindex = 0;
  int rightindex = 0;
  
  for (int i = left;i < right + 1;i++)
  {
    if (s_left_part > leftindex && s_right_part > rightindex)
    {
      if (leftpart[leftindex] > rightpart[rightindex])
      {
        arr[i] = rightpart[rightindex];
        rightindex += 1;
      }
      else
      {
        arr[i] = leftpart[leftindex];
        leftindex += 1;
      }
    }
    else if (s_left_part > leftindex) 
    {
     arr[i] = leftpart[leftindex]; 
     leftindex+= 1 ;
    }
    else
    {
     arr[i] = rightpart[rightindex]; 
     rightindex+= 1 ;
    }
  }
}

void merge_sort_algorithm(int arr[],int left,int right)
{
  if (left < right)
  {
    int mid = (left + right) /2;
    merge_sort_algorithm(arr, left,mid);
    merge_sort_algorithm(arr, mid+1,right);
    merge(arr,left,mid,right);
  }
}


int main()
{
  int arr[] = {9,6,4,8,6,9};
  int size = sizeof(arr) / sizeof(arr[0]);
  
  merge_sort_algorithm(arr,0,size -1);
  
  for (int k = 0; k < size;k++)
  {
    printf("arr %d = %d\n",k,arr[k]);
  }

  return 0;
}
